### ISSUE TITLE
Apply correct SQL condition for the QGIS Browser PG Layer Properties Dialog

### ISSUE THREAD

**Issue by palmerj** [qgis/QGIS#32276]
Checklist before submitting

- [x] Search through existing issue reports and gis.stackexchange.com to check whether the issue already exists
- [x] Test with a [clean new user profile](https://docs.qgis.org/testing/en/docs/user_manual/introduction/qgis_configuration.html?highlight=profile#working-with-user-profiles).
- [ ] Create a light and self-contained sample dataset and project file which demonstrates the issue

If the issue concerns a **third party plugin**, then it **cannot** be fixed by the QGIS team. Please raise your issue in the dedicated bug tracker for that specific plugin (as listed in the plugin's description). -->

**Describe the bug**

When you right click and select the layer properties for a QGIS PG table browser layer that contains multiple base geometry types (e.g one table but shows up multiple times in the browser for each polygon/line/point type) the SRID and Geometry Type query is called without the geometry type condition . The count and extents queries do correctly apply this filter. 

Likely the following SQL needs to be appended to the query

```sql

WHERE upper(geometrytype("the_geom")) IN xxxx
````

But I don't understand why the SRID and Geometry Types are required for this dialogue as they have already been determined to populated the Browser nodes. Maybe this query is not needed at all? 

**QGIS and OS versions**
QGIS version | 3.9.0-Master | QGIS code revision | 5f24a0b214
-- | -- | -- | --
Compiled against Qt | 5.12.3 | Running against Qt | 5.12.3
Compiled against GDAL/OGR | 2.4.1 | Running against GDAL/OGR | 2.4.1
Compiled against GEOS | 3.7.2-CAPI-1.11.2 | Running against GEOS | 3.7.2-CAPI-1.11.2 b55d2125
Compiled against SQLite | 3.28.0 | Running against SQLite | 3.28.0
PostgreSQL Client Version | 11.3 | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.1
Compiled against PROJ | 5.2.0 | Running against PROJ | Rel. 5.2.0, September 15th, 2018
OS Version | macOS 10.15
Active python plugins | QuickWKT; qgisnetworklogger; Mergin; processing; db_manager; MetaSearch

**Comment by strk** [comment]
It looks like getting layer properties is implemented creating a temporary layer:

> src/gui/qgsbrowserdockwidget_p.cpp:180 : (setItem) [1ms] creating temporary layer using path pg:/:@5432/public/t

The URI used does not include geometry column at all

**Comment by strk** [comment]
The code says:
```
  // find root item
  // we need to create a temporary layer to get metadata
  // we could use a provider but the metadata is not as complete and "pretty"  and this is easier
  QgsDebugMsg( QStringLiteral( "creating temporary layer using path %1" ).arg( layerItem->path() ) );
```

I'm not sure what `we could use a provider` means, but for sure "this is easier" reveals this can be improved :)

**Comment by strk** [comment]
The browser function takes a QgsDataItem and static-casts to QgsLayerItem. The Postgres provider's implementation of QgsLayerItem does contain a QgsPostgresLayerProperty member, so all the info needed by Browser are indeed available. It's only a matter of figuring out what's a good way to expose them, trough the new API. @elpaso do you think we could have an abstract QgsLayerProperty class ?

**Comment by elpaso** [comment]
Do you mean `QVariantMap QgsAbstractDatabaseProviderConnection::TableProperty::info()` ?
or is too generic?
How would `QgsLayerProperty` look like?

**Comment by elpaso** [comment]
See also: `QgsAbstractDatabaseProviderConnection::TableProperty`

**Comment by strk** [comment]
What I mean is that browserdockwidget has a pointer to `QgsLayerItem` and needs to retrive those properties somehow, to show them (currently calls htmlProperties, which is a method of VectorLayer).

For now I tried to take a different approach: trust the request in the URI for type/srid rather than checking them again. @palmerj please see how you like #32292 

Failure to detect geometry type via constraint (even when an explicit type was requested by user, before my PR) ends up retriving _all_ layer types again, which is another thing that could be improved...

**Comment by palmerj** [comment]
@strk. The I still don't see the correct filters being applied to the table and it's returning the full table count and extent etc. He's a snapshot of SQL queries that only get run when I load the layer property dialog:

|query                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |calls|total_time|rows|
|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----|----------|----|
|SELECT indexrelid FROM pg_index WHERE indrelid=?::regclass AND (indisprimary OR indisunique) ORDER BY CASE WHEN indisprimary THEN ? ELSE ? END LIMIT ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |1    |0.02      |1   |
|SELECT pg_is_in_recovery()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |1    |0.005     |1   |
|SELECT count(*) FROM pg_inherits WHERE inhparent=?::regclass                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |1    |0.008     |1   |
|DECLARE oidcursor BINARY CURSOR FOR select regclass('pg_class')::oid                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |2    |0.028     |0   |
|CLOSE oidcursor                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |2    |0.021     |0   |
|SET application_name='QGIS'                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |2    |0.009     |0   |
|SELECT attname,attnotnull FROM pg_index,pg_attribute WHERE indexrelid=? AND indrelid=attrelid AND pg_attribute.attnum=any(pg_index.indkey)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |1    |0.019     |1   |
|BEGIN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |2    |0.002     |0   |
|SELECT upper(type),srid,coord_dimension FROM geometry_columns WHERE f_table_name=? AND f_geometry_column=? AND f_table_schema=?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |1    |0.126     |1   |
|SELECT has_table_privilege(?,?),has_any_column_privilege(?,?),has_column_privilege(?,?,?),has_table_privilege(?,?),current_schema()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |1    |0.064     |1   |
|SELECT postgis_geos_version(),postgis_proj_version()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |2    |0.016     |2   |
|SELECT oid,typname,typtype,typelem,typlen FROM pg_type                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |1    |4.898     |5695|
|select description from pg_catalog.pg_description where objoid = $1 and classoid = ?::pg_catalog.regclass and objsubid = $2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |22   |0.042     |0   |
|SELECT count(*) FROM "bde"."crs_parcel"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |1    |855.317   |1   |
|set client_encoding to 'UNICODE'                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |2    |0.027     |0   |
|SELECT description FROM pg_description WHERE objoid=? AND objsubid=?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |1    |0.024     |1   |
|SELECT oid FROM pg_catalog.pg_extension WHERE extname = ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |2    |0.03      |0   |
|COMMIT                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |2    |0.002     |0   |
|SELECT query, calls, total_time, rows FROM pg_stat_statements;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |1    |0.144     |30  |
|SELECT pg_catalog.quote_ident(c.relname) FROM pg_catalog.pg_class c WHERE c.relkind IN (?, ?, ?, ?, ?, ?) AND substring(pg_catalog.quote_ident(c.relname),?,?)=? AND pg_catalog.pg_table_is_visible(c.oid) UNION SELECT pg_catalog.quote_ident(n.nspname) &#124;&#124; ? FROM pg_catalog.pg_namespace n WHERE substring(pg_catalog.quote_ident(n.nspname) &#124;&#124; ?,?,?)=? AND (SELECT pg_catalog.count(*) FROM pg_catalog.pg_namespace WHERE substring(pg_catalog.quote_ident(nspname) &#124;&#124; ?,?,?) = substring(?,?,pg_catalog.length(pg_catalog.quote_ident(nspname))+?)) > ? UNION SELECT pg_catalog.quote_ident(n.nspname) &#124;&#124; ? &#124;&#124; pg_catalog.quote_ident(c.relname) FROM pg_catalog.pg_class c, pg_catalog.pg_namespace n WHERE c.relnamespace = n.oid AND c.relkind IN (?, ?, ?, ?, ?, ?) AND substring(pg_catalog.quote_ident(n.nspname) &#124;&#124; ? &#124;&#124; pg_catalog.quote_ident(c.relname),?,?)=? AND substring(pg_catalog.quote_ident(n.nspname) &#124;&#124; ?,?,?) = substring(?,?,pg_catalog.length(pg_catalog.quote_ident(n.nspname))+?) AND (SELECT pg_catalog.count(*) FROM pg_catalo|11   |37.734    |2   |
|SELECT COUNT(*) FROM information_schema.tables WHERE table_name=?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |2    |5.315     |2   |
|SELECT pg_stat_statements_reset();                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |1    |0.031     |1   |
|SELECT * FROM "bde"."crs_parcel" LIMIT ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |2    |0.081     |1   |
|SELECT postgis_version()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |2    |0.01      |2   |
|fetch forward 1 from oidcursor                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |2    |0.01      |0   |
|SELECT attrelid, attnum, pg_catalog.format_type(atttypid,atttypmod), pg_catalog.col_description(attrelid,attnum), pg_catalog.pg_get_expr(adbin,adrelid), atttypid, attnotnull::int, indisunique::int FROM pg_attribute LEFT OUTER JOIN pg_attrdef ON attrelid=adrelid AND attnum=adnum LEFT OUTER JOIN ( SELECT DISTINCT indrelid, indkey, indisunique FROM pg_index WHERE indisunique ) uniq ON attrelid=indrelid AND attnum::text=indkey::text  WHERE attrelid IN (?)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |1    |0.482     |22  |
|SELECT oid FROM pg_catalog.pg_type WHERE typname=?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |2    |0.029     |2   |
|SELECT relkind FROM pg_class WHERE oid=regclass(?)::oid                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |1    |0.011     |1   |
|SELECT st_extent("shape") FROM "bde"."crs_parcel"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |1    |3334.386  |1   |
|select regclass(?)::oid                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |3    |0.026     |3   |
|SELECT ? FROM pg_class,pg_namespace WHERE pg_class.relnamespace=pg_namespace.oid AND pg_has_role(relowner,?) AND relname=? AND nspname=?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |1    |0.056     |1   |
|SELECT EXISTS ( SELECT c.oid FROM pg_class AS c JOIN pg_namespace AS n ON c.relnamespace=n.oid WHERE n.nspname=? AND c.relname=? )                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |2    |0.073     |2   |

Interesting that there are still so many queries (but aligns with your comment that a new temp layer is setup to get the details)  and `select description from pg_catalog.pg_description` is run 22 times!

**Comment by strk** [comment]
In order to avoid mixing different issues, let's try to keep this one focused on the title ("correct SQL condition"). The original report was about unused conditions when fetching types and srids: "SRID and Geometry Type query is called without the geometry type condition". My PR #32292 completely _drops_ those queries when a SRID and a Type is requested via the URI (this is what happens when Browser creates a temporary layer).

If that part is confirmed as a fix, and the behavior is acceptable, I'd consider the PR good so far.

Now, what you add here is that there are a couple more queries which are lacking a filter:

  - Extent query (`SELECT st_extent("shape") FROM "bde"."crs_parcel"`)
  - Count query (`SELECT count(*) FROM "bde"."crs_parcel"`)

As those are also good candidates for filtering by type/SRID I'll see if I can handle to add the the filter as part of the same PR.

For the other issue of further reducing round-trips (pg_description query) I'd ask to file another issue.

Please add further comments about changes related to type/srid filtering to the PR directly (#32292)

**Comment by strk** [comment]
@palmerj with #32292 merged I think we can close this issue, do you confirm correct SQL conditions are now set for all queries ?

**Comment by strk** [comment]
Assuming this was fixed by #32292
